### PR TITLE
Java client

### PR DESCRIPTION
**Pull request by philrykoff**
Hi all,

as announced in vertx-dev, this PR implements the following eventbus-bridge-clients tickets:

- Add configuration: pingInterval https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/9
- Add configuration: Path at which the WebSocket listener listens on the host https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/10
- Remove message response handlers from consumerMap after DeliveryOptions.timeout https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/11
- EventBusClient.DEFAULT_OPTIONS should not be static; add setter https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/12
- Reconnection handling https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/13
- EventBusClient configuration https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/14 (includes TLS support)
- Add proxy support https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/16
- Add logging https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/17

Added on 29th of October based on review comments and other impulses:

- Restore Java 7 compatibility https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/20
- Don't register message reply addresses at server https://github.com/vert-x3/vertx-eventbus-bridge-clients/issues/21

As this is my first PR and feels like a rather big one, I am most grateful for any review comments. I expect there to be some todos from those comments, which is why i created this PR before finishing 100% - it wouldn't be 100% anyway. 

Todos:
- [x]  Check & improve class / method source documentation
- [x]  Finish proxy implementation & tests. Commit.
- [x] Implement TcpBus tests with SOCKS Proxy (which will also be run for all other tests based on that)
- [ ] integration tests (with failsafe) that run the bridge without some classes: without jackson and with gson
- [x]  Regression test and bugfixing

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.vertx</groupId>
+    <artifactId>vertx-parent</artifactId>
+    <version>10</version>
+  </parent>
+
+  <properties>
+    <netty.version>4.1.8.Final</netty.version>
```
Comment (vietj) [inline java/pom.xml]: Rather Netty 4.1.15.Final now

File: java/pom.xml [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+      <version>2.7.4</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>io.vertx</groupId>
+      <artifactId>vertx-core</artifactId>
+      <version>3.4.2-SNAPSHOT</version>
```
Comment (vietj) [inline java/pom.xml]: rather 3.5.0 now

File: java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  </dependencies>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-compiler-plugin</artifactId>
+          <configuration>
+            <source>1.8</source>
```
Comment (vietj) [inline java/pom.xml]: should be 1.7 and not 1.8 for source/target

File: java/src/main/java/io/vertx/ext/eventbus/client/AsyncResult.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class AsyncResult<T> {
+
+  private final T result;
+  private final Throwable cause;
+
+  public AsyncResult(T result, Throwable cause) {
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/AsyncResult.java]: make it private and make two static method to build it : success(T) and failure(Throwable)

File: java/src/main/java/io/vertx/ext/eventbus/client/EventBusClientOptions.java [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+   * The default value of the pause between reconnect tries = 3000 ms
+   */
+  public static final int DEFAULT_AUTO_RECONNECT_INTERVAL = 3000;
+
+  /**
+   * The default value of the maximum number of auto reconnect tries = 0 (0 = no limit)
+   */
+  public static final int DEFAULT_MAX_AUTO_RECONNECT_TRIES = 0;
+
+  private TcpTransportOptions tcpTransportOptions;
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/EventBusClientOptions.java]: we should have one field transportOptions and not two that is TransportOptions

File: java/src/main/java/io/vertx/ext/eventbus/client/logging/Logger.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package io.vertx.ext.eventbus.client.logging;
+
+public interface Logger {
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/logging/Logger.java]: logging in general should not be public and should go in an impl package

File: java/src/main/java/io/vertx/ext/eventbus/client/options/PemTrustOptions.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+
+/**
+ * @author <a href="mailto:pl@linux.com">Phil Lehmann</a>
+ */
+public class PemTrustOptions extends TrustOptions {
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/options/PemTrustOptions.java]: so for PemTrust I think that netty provide an implementation (I didn't know it when I implemented it in Vert.x)

File: java/src/main/java/io/vertx/ext/eventbus/client/options/ProxyOptions.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
+package io.vertx.ext.eventbus.client.options;
+
+import java.util.Objects;
+
+/**
+ * @author <a href="mailto:pl@linux.com">Phil Lehmann</a>
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/options/ProxyOptions.java]: are you the brother of Alex btw :-) ?

File: java/src/main/java/io/vertx/ext/eventbus/client/transport/Transport.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+import java.security.KeyStore;
+import java.security.SecureRandom;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public abstract class Transport extends ChannelInitializer {
+
+  protected EventBusClientOptions options;
```
Comment (vietj) [inline java/src/main/java/io/vertx/ext/eventbus/client/transport/Transport.java]: relevant options field should be inline in the class instead and made final.

**Comment by vietj** [comment]
we need tests:
- for proxy
- we need integration tests (with failsafe) that run the bridge without some classes: without jackson and with gson, to check it works when the lib is not present. Failsafe allows to remove some dependency from classpath

File: java/src/main/java/io/vertx/ext/eventbus/client/options/ProxyOptions.java (hunk 8 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/options/ProxyOptions.java]: Nooo, where I come from every second guy is called Lehmann ;)

File: java/src/main/java/io/vertx/ext/eventbus/client/EventBusClientOptions.java (hunk 5 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/EventBusClientOptions.java]: Thanks for your review! Your comments seem to be straighforward, I'll get to it.

File: java/src/main/java/io/vertx/ext/eventbus/client/AsyncResult.java (hunk 4 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/AsyncResult.java]: Static factory methods for generic types look a bit clunky - I hope this is what you intended it to be :)

File: java/src/main/java/io/vertx/ext/eventbus/client/transport/Transport.java (hunk 9 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/transport/Transport.java]: I made it final, but could you elaborate what you mean with the first part?

File: java/src/main/java/io/vertx/ext/eventbus/client/logging/Logger.java (hunk 6 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/logging/Logger.java]: I created an impl package. Yet, not all classes could be made package private as `JULLoggerFactory` within `impl` is being referenced from `LoggerFactory` (parent ns) and `LoggerFactoryImplBase` needs to be accessed by the other namespace (wherever it is located), too.

Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/logging/Logger.java]: If all logging classes would reside in the same package (no `impl` package), I could make those two private.

**Comment by philrykoff** [comment]
There were already two tests for HTTP proxies in WebSocketBusTest (`testProxyHttp` and `testProxyHttpFailure`), so you were probably thinking about `TcpBusTest` SOCKS proxy tests? I just found https://github.com/ravn/jsocks, will add two similar tests for `TcpBusTest`.

File: java/src/main/java/io/vertx/ext/eventbus/client/options/PemTrustOptions.java (hunk 7 above)
Comment (philrykoff) [reply java/src/main/java/io/vertx/ext/eventbus/client/options/PemTrustOptions.java]: What I found was `PemPrivateKey` which was added for netty servers (https://github.com/netty/netty/pull/5377). I could find no class for PEM certs, which is pretty similar though.

**Comment by philrykoff** [comment]
As ravn/jsocks is not available via maven, I added the maven central & jitpack repositories, to fetch it via jitpack from github directly.

**Comment by cescoffier** [comment]
This may cause an issue for the release to maven central. I think there is a requirement that everything is resolved from central.

**Comment by philrykoff** [comment]
While I don't have any experience on publishing to central, their req page (http://central.sonatype.org/pages/requirements.html) states "In addition we discourage the usage of <repositories> and <pluginRepositories> and instead publish any required components to the Central Repository.".

So while it does not sound like a hard requirement, they are completely right to discourage that (though we would only need it for a test scope dependency). Solutions might be to use another SOCKS proxy server (I could only find mirrors of that one, no other solutions) or to mirror it in `vert-x3` and release it to maven ourselves. What do you think?

**Comment by philrykoff** [comment]
Filed https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/31, as pings seem to be broken for tcp bridge.

**Comment by philrykoff** [comment]
VertX core implements it's own proxies (does not use third party libs) for testing vertx-web proxy functionality. We I can get rid of ravn/jsocks (the jitpack dependancy), but for that to work we need this PR merged https://github.com/eclipse/vert.x/pull/2189 to enable SOCKS functionality for target IPs (at least on my machine, netty always translates hostnames to IPs before sending them to the proxy).

File: java/pom.xml (hunk 1 above)
Comment (philrykoff) [reply java/pom.xml]: Updated

File: java/pom.xml (hunk 2 above)
Comment (philrykoff) [reply java/pom.xml]: Done

File: java/pom.xml (hunk 3 above)
Comment (philrykoff) [reply java/pom.xml]: Done.

**Comment by theoweiss** [comment]
Are there any plans when to proceed?
As soon as this PR is merged I could provide a PR adding KeyCertOptions for tls client certificates.

**Comment by philrykoff** [comment]
Well, my PR https://github.com/eclipse/vert.x/pull/2189 would need to be merged as a foundation for this one. 

I did not yet have time to test if the fix to https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/31 worked.

Julien recommended to add "integration tests (with failsafe) that run the bridge without some classes: without jackson and with gson, to check it works when the lib is not present. Failsafe allows to remove some dependency from classpath", which I have no experience in.

Other than that, all review comments have been adressed imho.

**Comment by theoweiss** [comment]
Maybe we should ping @vietj how to proceed?

**Comment by vietj** [comment]
can you try with 3.5.1-SNAPSHOT and see if it works with the https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/31 patch ?

**Comment by philrykoff** [comment]
I‘ll do it within the next 2 days.

**Comment by vietj** [comment]
I need to review / merge the IPV4 support in core i guess

**Comment by philrykoff** [comment]
I have tested against 3.5.1-SNAPSHOT with the following result:

- TcpBusTest/testPing yields a warning, but works (is addressed with PR https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/31)
- TcpBusTest/testUnsubscribe fails (is addressed with https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/34)
- *BusTest/testProxySocks5* fail (is addressed with https://github.com/eclipse/vert.x/pull/2189)

All other tests succeed.

**Comment by philrykoff** [comment]
As soon as those get merged, I'll retest.

**Comment by theoweiss** [comment]
This is a really long running PR. Is there a chance that this is finished / merged before 3.6.0?

**Comment by vietj** [comment]
@philrykoff any news about the progress of this effort ?

**Comment by philrykoff** [comment]
The only issue as of now is that TcpBusTest/testUnsubscribe fails. This is a server side issue addressed with vert-x3/vertx-tcp-eventbus-bridge#34. As soon as that PR get's merged, I cloud retest this PR again.

**Comment by philrykoff** [comment]
Still waiting for vert-x3/vertx-tcp-eventbus-bridge#34.

**Comment by andreas-eberle** [comment]
@philrykoff: If I saw that right, vert-x3/vertx-tcp-eventbus-bridge#34 should have been fixed with vert-x3/vertx-tcp-eventbus-bridge#39. 

Could this be finished then? I really need this for my project. Is there a nightly version on Maven Central or so already?

**Comment by theoweiss** [comment]
@philrykoff @vietj Can we expect any progress here?

**Comment by philrykoff** [comment]
> @philrykoff: If I saw that right, [vert-x3/vertx-tcp-eventbus-bridge#34](https://github.com/vert-x3/vertx-tcp-eventbus-bridge/issues/34) should have been fixed with [vert-x3/vertx-tcp-eventbus-bridge#39](https://github.com/vert-x3/vertx-tcp-eventbus-bridge/pull/39).
> 
> Could this be finished then? I really need this for my project. Is there a nightly version on Maven Central or so already?

Somehow missed these comments in July. The last issue was fixed with vert-x3/vertx-tcp-eventbus-bridge#39 indeed, so from my point this can be merged.

**Comment by vietj** [comment]
I will check the branch out and give it a spin

**Comment by vietj** [comment]
@philrykoff I will try to simplify the code, specially the configuration part which has many options (similar to vertx core) which aren't necessary here

**Comment by vietj** [comment]
I've simplified the PR (streamlined options and reused some of Netty's class) and updated the branch here https://github.com/vert-x3/vertx-eventbus-bridge-clients/tree/java-client

**Comment by vietj** [comment]
I think we need documentation and examples

**Comment by vietj** [comment]
Closing this in favor of https://github.com/vert-x3/vertx-eventbus-bridge-clients/pull/24

**Comment by vietj** [comment]
merged here https://github.com/vert-x3/vertx-eventbus-bridge-clients/pull/24

it took long time to finish thanks to @gaol - I want to also thank @PhilLehmann for putting lot of efforts in this

**Pull request closed without merging**